jobs/src/main/scala/dcos/metronome/jobrun/impl/JobRunExecutorActor.scala

```diff
@@ -55,17 +55,20 @@ class JobRunExecutorActor(
       case JobRunStatus.Initial =>
         becomeCreating()
       case JobRunStatus.Starting | JobRunStatus.Active =>
-        val existingTasks = tasksFromInstanceTracker()
-        existingTasks match {
-          case tasks if tasks.nonEmpty =>
-            // the actor was probably just restarted and we did not lose contents of the launch queue
-            self ! Initialized(tasks)
-          case _ if jobRun.status == JobRunStatus.Starting && existsInLaunchQueue() =>
-            // we did attempt to launch something but there is no known existing task yet
-            // this is exactly the same as starting state with already launched item
-            context.become(starting)
-          case _ =>
-            self ! Initialized(Nil)
+        async {
+          val existingTasks = await(tasksFromInstanceTracker())
+          val alreadyQueued = await(existsInLaunchQueue())
+          existingTasks match {
+            case tasks if tasks.nonEmpty =>
+              // the actor was probably just restarted and we did not lose contents of the launch queue
+              self ! Initialized(tasks)
```

Comment:
I think in reality this is fine; having async code access the actor context is error prone, however. We should ideally switch this to pipeTo. I'd also prefer to factor out the async logic to an object method so it does not have access to the internal actor state.

Reply:
I extracted out the async block to an object method